crates/iceberg-ext/src/spec/spec_binder.rs

```diff
@@ -0,0 +1,476 @@
+use crate::catalog::rest::ErrorModel;
+use http::StatusCode;
+use iceberg::spec::{
+    NestedFieldRef, PartitionField, PartitionSpec, SchemaRef, Transform, Type,
+    UnboundPartitionField, UnboundPartitionSpec,
+};
+use std::cmp::max;
+use std::collections::HashSet;
+
+type Result<T> = std::result::Result<T, ErrorModel>;
+
+#[derive(Debug)]
+pub struct PartitionSpecBinder {
+    spec_id: i32,
+    schema: SchemaRef,
+    partition_names: HashSet<String>,
+    last_assigned_field_id: i32,
+}
+
+impl PartitionSpecBinder {
+    const PARTITION_DATA_ID_START: i32 = 1000;
+
+    pub fn new(schema: SchemaRef, spec_id: i32) -> Self {
+        Self {
+            spec_id,
+            schema,
+            last_assigned_field_id: Self::PARTITION_DATA_ID_START - 1,
+            partition_names: HashSet::new(),
+        }
+    }
+
+    pub fn bind(mut self, spec: &UnboundPartitionSpec) -> Result<PartitionSpec> {
+        Ok(PartitionSpec {
+            spec_id: self.spec_id,
+            fields: spec
+                .fields
+                .iter()
+                .map(|unbounded_field| self.bind_field(unbounded_field))
+                .collect::<Result<Vec<PartitionField>>>()?,
+        })
+    }
+
+    fn bind_field(&mut self, spec_field: &UnboundPartitionField) -> Result<PartitionField> {
+        self.check_partition_names(&spec_field.name)?;
+
+        let schema_field = self.get_schema_field_by_name(&spec_field.name)?;
+
+        self.check_transform_compatibility(&spec_field.transform, &schema_field.field_type)?;
+        Self::check_schema_field_eq_source_id(schema_field, spec_field.source_id)?;
+
+        self.update_names(&spec_field.name);
+
+        let partition_id = spec_field
+            .partition_id
+            .unwrap_or_else(|| self.increment_and_get_next_partition_id());
```

Comment:
I was really confused by the naming of the `partition_id` attribute - which is a really bad name in the core rust crate.
I would prefer to rename `increment_and_get_next_partition_id` to `increment_and_get_next_field_id` (since its modifying `self.last_assigned_*field*_id`) and also name the assigned variable let field_id instead of partition_id.
Otherwise line 64 is also quite confusing where we take the max between a field and a partition id.
We could then add a small comment in line 52 stating that the partition_id attribute is not actually the id of this partition, but instead the identifier of the field in the partition. What do you think?